marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+
+
+## Representation of Marlowe Semantics in Plutus
+
+Marlowe semantics are embodied in the Haskell function `computeTransaction` that is generated from [the Isabelle implementation of the Marlowe DSL](https://github.com/input-output-hk/marlowe/tree/master/isabelle/Core). Note that `computeTransaction`  uses the `PlutusTx` types and functions instead of the standard-prelude or GHC-base Haskell ones.
+```haskell
+computeTransaction :: TransactionInput -> State -> Contract -> TransactionOutput
+```
+
+With the exception of `Input`, [the Isabelle types for Marlowe](https://github.com/input-output-hk/marlowe/blob/master/isabelle/Core/SemanticsTypes.thy) directly correspond to their Haskell representations, so below we only provide the definitions that will be later referenced in this document.
+```haskell
+data TransactionInput =
+    TransactionInput
+    {
+      txInterval :: TimeInterval
+    , txInputs   :: [Input]
+    }
+
+type TimeInterval = (POSIXTime, POSIXTime)
+
+data State =
+    State
+    {
+      accounts    :: Accounts
+    , choices     :: Map ChoiceId ChosenNum
+    , boundValues :: Map ValueId Integer
+    , minTime     :: POSIXTime
+    }
+
+type Accounts = Map (AccountId, Token) Integer
+
+data TransactionOutput =
+    TransactionOutput
+    {
+      txOutWarnings :: [TransactionWarning]
+    , txOutPayments :: [Payment]
+    , txOutState    :: State
+    , txOutContract :: Contract
+    }
+  | Error TransactionError
+```
+
+The `Input` of the Isabelle definition corresponds to the `InputContent` of the Haskell representation because the latter includes support for merkleization of contract continuations in `Case` statements. Note that Merkleization is not currently treated by the Isabelle specification of Marlowe.
+```haskell
+data Input =
+    NormalInput InputContent
+  | MerkleizedInput InputContent BuiltinByteString Contract
+```
+In the above, the `BuiltinByteString` is the hash of the serialized continuation `Contract`.
+
+
+## Plutus Validator for Marlowe Semantics
+
+The Marlowe validator is an unparameterized interpreter of Marlowe semantics. Thus, the script address of the Marlowe validator is independent of the particular contract and roles currency.
+```haskell
+smallMarloweValidator :: MarloweData -> MarloweInput -> ScriptContext -> Bool
+```
+
+
+### Types
+
+The Plutus `Datum` for Marlowe simply bundles the role-currency and payout information with the `State` and the `Contract`:
+```haskell
+data MarloweData =
+    MarloweData
+    {
+      marloweParams   :: MarloweParams
+    , marloweState    :: State
+    , marloweContract :: Contract
+    }
+
+data MarloweParams =
+    MarloweParams
+    {
+      rolePayoutValidatorHash :: ValidatorHash
```

Comment:
Do we want to consider hard coding the `rolePayoutValidatorHash` (it is going to be constant) value in the validator itself? We can save some space in the `Tx` and save some decoding/encoding CPU cycles.

Reply:
Yes, I added a `FIXME` to address this after the validator is finalized.